### PR TITLE
Make CHIP devices advertise as operational devices via mDNS

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem

Need CHIP devices to respond to mDNS properly.

 #### Summary of Changes

 Replaces DiscoveryManager (which has avahi/esp32 implementation only) with an abstracted-away minMDNS implementation for advertising. Tested via:

```sh
andrei@Andreis-Mac-mini ~ % dns-sd -B _chip._tcp                 
Browsing for _chip._tcp
DATE: ---Tue 08 Dec 2020---
16:27:39.955  ...STARTING...
Timestamp     A/R    Flags  if Domain               Service Type         Instance Name
16:27:40.157  Add        2   4 local.               _chip._tcp.          14A77CBB3-BC5C01
^C
andrei@Andreis-Mac-mini ~ % dns-sd -L 14A77CBB3-BC5C01 _chip._tcp       
Lookup 14A77CBB3-BC5C01._chip._tcp.local
DATE: ---Tue 08 Dec 2020---
16:27:44.658  ...STARTING...
16:27:44.872  14A77CBB3-BC5C01._chip._tcp.local. can be reached at 14A77CBB3-BC5C01.local.:11097 (interface 4)
 =
^C
andrei@Andreis-Mac-mini ~ % dns-sd -G v4v6 14A77CBB3-BC5C01.local            
DATE: ---Tue 08 Dec 2020---
16:27:50.531  ...STARTING...
Timestamp     A/R    Flags if Hostname                               Address                                      TTL
16:27:50.752  Add        3  4 14A77CBB3-BC5C01.local.                FE80:0000:0000:0000:9AF4:ABFF:FE6B:1AFC%en0  30
16:27:50.752  Add        2  4 14A77CBB3-BC5C01.local.                10.0.0.200                                   30
^C
andrei@Andreis-Mac-mini ~ % ping6 FE80:0000:0000:0000:9AF4:ABFF:FE6B:1AFC%en0
PING6(56=40+8+8 bytes) fe80::1823:ac1e:caee:43c2%en0 --> fe80::9af4:abff:fe6b:1afc%en0
16 bytes from fe80::9af4:abff:fe6b:1afc%en0, icmp_seq=0 hlim=255 time=72.875 ms
16 bytes from fe80::9af4:abff:fe6b:1afc%en0, icmp_seq=1 hlim=255 time=92.139 ms
^C
--- FE80:0000:0000:0000:9AF4:ABFF:FE6B:1AFC%en0 ping6 statistics ---
2 packets transmitted, 2 packets received, 0.0% packet loss
round-trip min/avg/max/std-dev = 72.875/82.507/92.139/9.632 ms
andrei@Andreis-Mac-mini ~ % ping 10.0.0.200                                  
PING 10.0.0.200 (10.0.0.200): 56 data bytes
64 bytes from 10.0.0.200: icmp_seq=0 ttl=255 time=21.517 ms
64 bytes from 10.0.0.200: icmp_seq=1 ttl=255 time=35.868 ms
^C
--- 10.0.0.200 ping statistics ---
2 packets transmitted, 2 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 21.517/28.693/35.868/7.175 ms
```

### CODE REVIEW

**Comment by andy31415** [comment]
Had a few bugfixes: invalid && return for builders, created separatte packetbuffer copy for every broadcast packet.

**Comment by andy31415** [comment]
Now that I started using minMDNS beyond hardcoded examples, I expect I will want a few passes to make it a bit nicer (constructor addresses and settings seem hard to follow). 

For now I take this as 'this works' and then will iterate to add 'commisioning advertising' and make the thing a bit more readable.

In parallel we can build the platform-dependent mDNS version (avahi and esp32 if applicable). minMDNS tries hard to only use static memory and never dynamic, which has sideffects in how pretty the code is.

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+{
+public:
+    const char * c_str() const { return mBuffer; }
+
+    StringBuilder & Append(const char * s)
+    {
+        size_t available = kSize - mUsed;
+        size_t len       = strlen(s);
+
+        if (len > available)
```
Comment (rwalker-apple) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:47]: `len >= available`?  perhaps also add a test for this class that verifies no buffer overrun with `strlen(s) == kSize`?

Comment (andy31415) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:47]: This was supposed to be debug only. 
However realistically I believe such a 'build debug messages' would be a really nice support class. I will extract it and add tests.

pigweed has stringbuilder however it is C++17 I think (uses string view). Shame.

Comment (rwalker-apple) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:47]: or build this class with `BufferWriter` and reserve the last byte for `\0` ?

File: src/lib/mdns/minimal/Server.cpp [hunk 2]
```diff
@@ -61,6 +61,26 @@ class ShutdownOnError
     ServerBase * mServer;
 };
 
+chip::System::PacketBufferHandle Clone(chip::System::PacketBufferHandle & original)
```
Comment (rwalker-apple) [inline src/lib/mdns/minimal/Server.cpp:64]: can you move this into the PacketBuffer class instead?

**Review by rwalker-apple — CHANGES_REQUESTED** [review]

File: src/lib/mdns/minimal/Server.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/mdns/minimal/Server.cpp:64]: I will move after I double-check if it is really needed. Had crashes before this, however I wonder if the crash was related to some packet buffer reference returns.

**Review by gjc13 — APPROVED** [review]

File: examples/common/chip-app-server/Server.cpp [hunk 3]
```diff
@@ -158,11 +159,30 @@ void InitServer(AppDelegate * delegate)
         SuccessOrExit(err = gRendezvousServer.Init(params, &gTransports));
     }
 
+    // TODO: advertise this only when really operational once we support both
+    // operational and commisioning advertising is supported.
+    {
+        constexpr uint64_t kTestFabricId = 5544332211;
+        err                              = Mdns::ServiceAdvertiser::Instance().Advertise(Mdns::OperationalAdvertisingParameters()
```
Comment (gjc13) [inline examples/common/chip-app-server/Server.cpp:166]: We can still guard with `IsFullyProvisioned` to support advertising `_chipc._udp` for IP rendezvous.

**Review by gjc13 — COMMENTED** [review]
Generally LGTM though I'd suggest to implement the mDNS platform API rather than introducing a new class.

File: examples/common/chip-app-server/Server.cpp (hunk 3 above)
Comment (andy31415) [reply examples/common/chip-app-server/Server.cpp:166]: Agree. The reason I made it advertise fully here is to always have functional mDNS to test with.

My next PRs (maybe pending some cleanup in usage) would be to add _chipc._udp vs operational advertisiement.

File: src/lib/mdns/minimal/Server.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/mdns/minimal/Server.cpp]: Clone is needed. Moved it and added a comment that we waste one clone in the code and that could be resolved if we care enough (we may not, since mDNS is expensive as is and not as frequent since we use unicast for most purposes when clone is not needed)

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp (hunk 1 above)
Comment (andy31415) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp]: Created a StringBuilder that uses a buffer writer and includes unit tests.

**Review by BroderickCarlin — APPROVED** [review]

File: src/lib/support/StringBuilder.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        NullTerminate();
+        return *this;
+    }
+
+    /// Append an integer value
+    StringBuilderBase & Add(int value)
+    {
+        char buff[32];
+        snprintf(buff, sizeof(buff), "%d", value);
+        buff[sizeof(buff) - 1] = 0;
```
Comment (rwalker-apple) [inline src/lib/support/StringBuilder.h:49]: this one isn't necessary

also the largest `int value` should consume less than 32 bytes :-)

**Review by rwalker-apple — APPROVED** [review]

File: src/system/SystemPacketBuffer.h [hunk 5]
```diff
@@ -612,6 +612,15 @@ class DLL_EXPORT PacketBufferHandle
...
     PacketBuffer * Get_ForNow() const { return mBuffer; }
 
+    /**
+     * Creates a copy of the data in this packet.
+     *
+     * Does NOT support chained buffers.
+     *
+     * @returns empty handle on allocation failure.
+     */
+    PacketBufferHandle Clone();
```
Comment (bzbarsky-apple) [inline src/system/SystemPacketBuffer.h:622]: Maybe call this `CloneData`?

File: src/lib/support/StringBuilder.h (hunk 4 above)
Comment (andy31415) [reply src/lib/support/StringBuilder.h:49]: Right ... even 64-bit is 20 chars at most. Wanted to not think about it since I imagine this code will be copied and pasted as if we ever want to format additional value types in here.

Do you think it would be useful to  drop the line or are ok to keep as is?

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    {
+        kIpV4,
+        kIpV6,
+    };
+    State mState;
+    chip::Inet::InterfaceIterator mIterator;
+
+    void SkipToFirstValidInterface()
+    {
+        if (SkipCurrentInterface())
```
Comment (mspang) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:131]:
```
do {
  if (!SkipCurrentInterface()) {
    break;
  }
} while (mIterator.Next());

```
?

File: src/lib/mdns/BUILD.gn [hunk 7]
```diff
@@ -14,6 +14,11 @@
 
 import("//build_overrides/chip.gni")
 
+declare_args() {
+  # Set up what advertiser to use for mDNS advertisement
+  chip_mdns_avertiser = "minimal-mdns"
```
Comment (mspang) [inline src/lib/mdns/BUILD.gn:19]: Suggest just "minimal"

File: src/system/SystemPacketBuffer.cpp [hunk 8]
```diff
@@ -536,5 +536,31 @@ PacketBufferHandle PacketBufferHandle::PopHead()
     return PacketBufferHandle(head);
 }
 
+PacketBufferHandle PacketBufferHandle::CloneData()
+{
+    if (!mBuffer->Next().IsNull())
```
Comment (mspang) [inline src/system/SystemPacketBuffer.cpp:541]: Could this be an assertion?

**Review by mspang — APPROVED** [review]

File: src/system/SystemPacketBuffer.cpp (hunk 8 above)
Comment (andy31415) [reply src/system/SystemPacketBuffer.cpp:541]: I am unsure how assertions work. nlASSERT?

Comment (andy31415) [reply src/system/SystemPacketBuffer.cpp:541]: All asserts in this file seem to be static. nlASSERT seem to be a noop in production (really? that seems odd, but this is what following #define seems to give me).

I would rather keep it as is as empty return.

**Comment by bzbarsky-apple** [comment]
@andy31415 Is the size of the codesize increase here about as expected?

**Comment by andy31415** [comment]
> @andy31415 Is the size of the codesize increase here about as expected?

I don't have much control on the sizes. Flash/txt seems ok, BSS size seems a bit large. I expect as we iterate we may find places where we can decrease things.
